File: backend/auth-service/src/main/java/com/englishflow/auth/service/InvitationService.java
package com.englishflow.auth.service;

import com.englishflow.auth.dto.AcceptInvitationRequest;
import com.englishflow.auth.dto.InvitationRequest;
import com.englishflow.auth.dto.InvitationResponse;
import com.englishflow.auth.entity.Invitation;
import com.englishflow.auth.entity.User;
import com.englishflow.auth.repository.InvitationRepository;
import com.englishflow.auth.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class InvitationService {

    private final InvitationRepository invitationRepository;
    private final UserRepository userRepository;
    private final EmailService emailService;
    private final PasswordEncoder passwordEncoder;

    @Transactional
    public InvitationResponse sendInvitation(InvitationRequest request, Long invitedBy) {
        // Validate role
        User.Role role;
        try {
            role = User.Role.valueOf(request.getRole().toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Invalid role: " + request.getRole());
        }

        // Only TUTOR and ACADEMIC_OFFICE_AFFAIR can be invited
        if (role != User.Role.TUTOR && role != User.Role.ACADEMIC_OFFICE_AFFAIR) {
            throw new IllegalArgumentException("Only TUTOR and ACADEMIC_OFFICE_AFFAIR roles can be invited");
        }

        // Check if email already exists
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new com.englishflow.auth.exception.EmailAlreadyExistsException(request.getEmail());
        }

        // Check if there's already a pending invitation
        if (invitationRepository.existsByEmailAndUsedFalse(request.getEmail())) {
            throw new IllegalArgumentException("An invitation has already been sent to this email");
        }

        // Create invitation
        String token = UUID.randomUUID().toString();
        Invitation invitation = Invitation.builder()
                .email(request.getEmail())
                .token(token)
                .role(role)
                .expiryDate(LocalDateTime.now().plusDays(7)) // 7 days to accept
                .used(false)
                .invitedBy(invitedBy)
                .build();

        Invitation savedInvitation = invitationRepository.save(invitation);

        // Send invitation email
        try {
            emailService.sendInvitationEmail(
                    request.getEmail(),
                    role.name(),
                    token
            );
            log.info("Invitation email sent to: {}", request.getEmail());
        } catch (Exception e) {
            log.error("Failed to send invitation email to: {}", request.getEmail(), e);
            // Don't throw exception - invitation is saved, can be resent
            log.warn("Invitation created but email failed. Can be resent later.");
        }

        return InvitationResponse.fromEntity(savedInvitation);
    }

    @Transactional(readOnly = true)
    public InvitationResponse getInvitationByToken(String token) {
        Invitation invitation = invitationRepository.findByToken(token)
                .orElseThrow(() -> new com.englishflow.auth.exception.InvalidTokenException("Invitation", "Token not found"));

        if (invitation.isUsed()) {
            throw new com.englishflow.auth.exception.InvitationAlreadyUsedException();
        }

        if (invitation.getExpiryDate().isBefore(LocalDateTime.now())) {
            throw new com.englishflow.auth.exception.InvitationExpiredException();
        }

        return InvitationResponse.fromEntity(invitation);
    }

    @Transactional
    public User acceptInvitation(AcceptInvitationRequest request) {
        // Find invitation
        Invitation invitation = invitationRepository.findByToken(request.getToken())
                .orElseThrow(() -> new com.englishflow.auth.exception.InvalidTokenException("Invitation", "Token not found"));

        // Validate invitation
        if (invitation.isUsed()) {
            throw new com.englishflow.auth.exception.InvitationAlreadyUsedException();
        }

        if (invitation.getExpiryDate().isBefore(LocalDateTime.now())) {
            throw new com.englishflow.auth.exception.InvitationExpiredException();
        }

        // Check if email already exists (double check)
        if (userRepository.existsByEmail(invitation.getEmail())) {
            throw new com.englishflow.auth.exception.EmailAlreadyExistsException(invitation.getEmail());
        }

        // Create user
        User user = new User();
        user.setEmail(invitation.getEmail());
        user.setFirstName(request.getFirstName());
        user.setLastName(request.getLastName());
        user.setPassword(passwordEncoder.encode(request.getPassword()));
        user.setPhone(request.getPhone());
        user.setCin(request.getCin());
        user.setDateOfBirth(request.getDateOfBirth());
        user.setAddress(request.getAddress());
        user.setCity(request.getCity());
        user.setPostalCode(request.getPostalCode());
        user.setBio(request.getBio());
        user.setYearsOfExperience(request.getYearsOfExperience());
        user.setRole(invitation.getRole());
        user.setActive(false); // Require admin activation
        user.setRegistrationFeePaid(false);
        user.setProfileCompleted(true);

        User savedUser = userRepository.save(user);

        // Mark invitation as used
        invitation.setUsed(true);
        invitation.setUsedAt(LocalDateTime.now());
        invitationRepository.save(invitation);

        // Send welcome email
        try {
            emailService.sendWelcomeEmail(savedUser.getEmail(), savedUser.getFirstName());
            log.info("Welcome email sent to: {}", savedUser.getEmail());
        } catch (Exception e) {
            log.error("Failed to send welcome email to: {}", savedUser.getEmail(), e);
        }

        return savedUser;
    }

    @Transactional(readOnly = true)
    public List<InvitationResponse> getAllInvitations() {
        return invitationRepository.findAll().stream()
                .map(InvitationResponse::fromEntity)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<InvitationResponse> getPendingInvitations() {
        return invitationRepository.findByUsedFalse().stream()
                .map(InvitationResponse::fromEntity)
                .collect(Collectors.toList());
    }

    @Transactional
    public void cancelInvitation(Long invitationId) {
        Invitation invitation = invitationRepository.findById(invitationId)
                .orElseThrow(() -> new com.englishflow.auth.exception.InvalidTokenException("Invitation", "Invitation not found"));

        if (invitation.isUsed()) {
            throw new com.englishflow.auth.exception.InvitationAlreadyUsedException();
        }

        invitationRepository.delete(invitation);
    }

    @Transactional
    public InvitationResponse resendInvitation(Long invitationId) {
        Invitation invitation = invitationRepository.findById(invitationId)
                .orElseThrow(() -> new com.englishflow.auth.exception.InvalidTokenException("Invitation", "Invitation not found"));

        if (invitation.isUsed()) {
            throw new com.englishflow.auth.exception.InvitationAlreadyUsedException();
        }

        // Extend expiry date
        invitation.setExpiryDate(LocalDateTime.now().plusDays(7));
        Invitation updatedInvitation = invitationRepository.save(invitation);

        // Resend email
        try {
            emailService.sendInvitationEmail(
                    invitation.getEmail(),
                    invitation.getRole().name(),
                    invitation.getToken()
            );
            log.info("Invitation email resent to: {}", invitation.getEmail());
        } catch (Exception e) {
            log.error("Failed to resend invitation email to: {}", invitation.getEmail(), e);
            // Don't throw - invitation is updated, email can be retried
            log.warn("Invitation updated but email failed. Can be resent again.");
        }

        return InvitationResponse.fromEntity(updatedInvitation);
    }

    @Transactional
    public void cleanupExpiredInvitations() {
        List<Invitation> expiredInvitations = invitationRepository
                .findByExpiryDateBeforeAndUsedFalse(LocalDateTime.now());
        
        if (!expiredInvitations.isEmpty()) {
            invitationRepository.deleteAll(expiredInvitations);
            log.info("Cleaned up {} expired invitations", expiredInvitations.size());
        }
    }
}

File: frontend/src/app/core/services/invitation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface InvitationRequest {
  email: string;
  role: 'TUTOR' | 'ACADEMIC_OFFICE_AFFAIR';
}

export interface InvitationResponse {
  id: number;
  email: string;
  role: string;
  expiryDate: string;
  used: boolean;
  invitedBy: number;
  createdAt: string;
  usedAt: string | null;
  token?: string; // Optional - only returned in some endpoints
}

export interface AcceptInvitationRequest {
  token: string;
  firstName: string;
  lastName: string;
  password: string;
  phone?: string;
  cin?: string;
  dateOfBirth?: string;
  address?: string;
  city?: string;
  postalCode?: string;
  bio?: string;
  yearsOfExperience?: number;
}

export interface AcceptInvitationResponse {
  token: string;
  id: number;
  email: string;
  firstName: string;
  lastName: string;
  role: string;
  profilePhoto: string | null;
  phone: string | null;
  profileCompleted: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class InvitationService {
  private apiUrl = `${environment.apiUrl}/auth/invitations`;

  constructor(private http: HttpClient) {}

  /**
   * Send an invitation to a user
   * @param request Invitation request with email and role
   */
  sendInvitation(request: InvitationRequest): Observable<InvitationResponse> {
    return this.http.post<InvitationResponse>(`${this.apiUrl}/send`, request);
  }

  /**
   * Verify an invitation token
   * @param token Invitation token from email
   */
  verifyInvitation(token: string): Observable<InvitationResponse> {
    return this.http.get<InvitationResponse>(`${this.apiUrl}/token/${token}`);
  }

  /**
   * Accept an invitation and create account
   * @param request Accept invitation request with user details
   */
  acceptInvitation(request: AcceptInvitationRequest): Observable<AcceptInvitationResponse> {
    return this.http.post<AcceptInvitationResponse>(`${this.apiUrl}/accept`, request);
  }

  /**
   * Get all invitations (admin only)
   */
  getAllInvitations(): Observable<InvitationResponse[]> {
    return this.http.get<InvitationResponse[]>(this.apiUrl);
  }

  /**
   * Get pending invitations (admin only)
   */
  getPendingInvitations(): Observable<InvitationResponse[]> {
    return this.http.get<InvitationResponse[]>(`${this.apiUrl}/pending`);
  }

  /**
   * Resend an invitation (admin only)
   * @param invitationId ID of the invitation to resend
   */
  resendInvitation(invitationId: number): Observable<InvitationResponse> {
    return this.http.post<InvitationResponse>(`${this.apiUrl}/${invitationId}/resend`, {});
  }

  /**
   * Cancel an invitation (admin only)
   * @param invitationId ID of the invitation to cancel
   */
  cancelInvitation(invitationId: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.apiUrl}/${invitationId}`);
  }

  /**
   * Cleanup expired invitations (admin only)
   */
  cleanupExpiredInvitations(): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(`${this.apiUrl}/cleanup`, {});
  }

  /**
   * Check if invitation is expired
   * @param expiryDate Expiry date string
   */
  isExpired(expiryDate: string): boolean {
    return new Date(expiryDate) < new Date();
  }

  /**
   * Get days until expiry
   * @param expiryDate Expiry date string
   */
  getDaysUntilExpiry(expiryDate: string): number {
    const expiry = new Date(expiryDate);
    const now = new Date();
    const diffTime = expiry.getTime() - now.getTime();
    const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
    return diffDays;
  }

  /**
   * Format role name for display
   * @param role Role string
   */
  formatRoleName(role: string): string {
    const roleNames: { [key: string]: string } = {
      'TUTOR': 'Tutor',
      'TEACHER': 'Teacher',
      'ACADEMIC_OFFICE_AFFAIR': 'Academic Affairs',
      'ADMIN': 'Administrator'
    };
    return roleNames[role] || role;
  }
}
